tutorials/partitioned-heat-conduction/fenicsx/heat.py

```diff
@@ -76,57 +77,89 @@ def determine_gradient(V_g, u, flux):
+V = FunctionSpace(mesh, scalar_element)
+V_g = FunctionSpace(mesh, vector_element)
 W = V_g.sub(0).collapse()
 
 # Define boundary conditions
-u_D = Expression('1 + x[0]*x[0] + alpha*x[1]*x[1] + beta*t', degree=2, alpha=alpha, beta=beta, t=0)
-u_D_function = interpolate(u_D, V)
+class Expression_u_D:
+    def __init__(self):
+        self.t = 0.0
+        self.alpha = alpha
+        self.beta = beta
+    def eval(self, x):
+        return np.full(x.shape[1], 1 + x[0]*x[0] + self.alpha*x[1]*x[1] + self.beta*self.t)
+
+u_D = Expression_u_D()
+u_D_function = Function(V)
+u_D_function.interpolate(u_D.eval)
 
 if problem is ProblemType.DIRICHLET:
     # Define flux in x direction
-    f_N = Expression("2 * x[0]", degree=1, alpha=alpha, t=0)
-    f_N_function = interpolate(f_N, W)
+    class Expression_f_N:
+        def __init__(self):
+            self.alpha = alpha
+            self.t = 0.0
+        def eval(self, x):
+            return np.full(x.shape[1], 2*x[0])
+
+    f_N = Expression_f_N()
+    f_N_function = Function(V)
+    f_N_function.interpolate(f_N.eval)
 
 # Define initial value
-u_n = interpolate(u_D, V)
-u_n.rename("Temperature", "")
+u_n = Function(V, name="Temperature")
+u_n.interpolate(u_D.eval)
+# u_n.rename("Temperature", "")
 
 precice, precice_dt, initial_data = None, 0.0, None
 
 # Initialize the adapter according to the specific participant
 if problem is ProblemType.DIRICHLET:
-    precice = Adapter(adapter_config_filename="precice-adapter-config-D.json")
+    precice = Adapter(MPI.COMM_WORLD, adapter_config_filename="precice-adapter-config-D.json")
     precice_dt = precice.initialize(coupling_boundary, read_function_space=V, write_object=f_N_function)
 elif problem is ProblemType.NEUMANN:
-    precice = Adapter(adapter_config_filename="precice-adapter-config-N.json")
+    precice = Adapter(MPI.COMM_WORLD, adapter_config_filename="precice-adapter-config-N.json")
     precice_dt = precice.initialize(coupling_boundary, read_function_space=W, write_object=u_D_function)
 
-dt = Constant(0)
-dt.assign(np.min([fenics_dt, precice_dt]))
+dt = np.min([fenics_dt, precice_dt])  # TODO does that work? This used to be a Constant() object in dolfin.
 
 # Define variational problem
 u = TrialFunction(V)
 v = TestFunction(V)
-f = Expression('beta - 2 - 2*alpha', degree=2, alpha=alpha, beta=beta, t=0)
-F = u * v / dt * dx + dot(grad(u), grad(v)) * dx - (u_n / dt + f) * v * dx
-
-bcs = [DirichletBC(V, u_D, remaining_boundary)]
+class Expression_f:
+    def __init__(self):
+        self.alpha = alpha
+        self.beta = beta
+        self.t = 0.0
+    def eval(self, x):
+        return np.full(x.shape[1], self.beta - 2 - 2*self.alpha)
+f = Expression_f()
+f_function = Function(V)
+f_function.interpolate(f.eval)
+F = u * v / dt * dx + dot(grad(u), grad(v)) * dx - (u_n / dt + f_function) * v * dx
+
+dofs_remaining = locate_dofs_geometrical(V, remaining_boundary)
+bcs = [DirichletBC(u_D_function, dofs_remaining)]
```

Comment:
Is this way of defining a `DirichletBC` allowed? If yes: Maybe we could somehow translate the `coupling_expression` below to something that "looks" like `u_D_function` to get it working?

Reply:
Yes, it is. Usually they are set with Function objects, see e.g. https://jorgensd.github.io/dolfinx-tutorial/chapter2/heat_code.html
It seems like the DirichletBC looks for ```coupling_expression.x.array```, which should by a numpy array (see https://github.com/FEniCS/dolfinx/blob/e0c3206281d71e0391482f4e07e3e30cc2cc1199/python/dolfinx/fem/dirichletbc.py#L137 ).